### DISCUSSION TITLE
Using UnoLib-CrossRuntime templated project(s) concurrently with an UnoApp templated project

### DISCUSSION THREAD

**Discussion by baskren** [unoplatform/uno#14854] [General]
I've spent a big chunk of time this week investigating the conversion of my collection of cross platform uno libraries from a terrible (but works) home grown collection of build files (.csproj, .props, .targets) to using the build files found in the UnoLib-CrossRuntime template as a foundation.  I ran into the below scenario [shown in this simple demo solution](https://github.com/unoplatform/uno/files/13686286/UnoApp1.zip).

1. Start with `dotnet new unoapp -preset=blank -o UnoApp1` templated solution
2. Add a `dotnet new unolib-crossruntime -o UnoLibCrossRuntime` library to the solution.  
3. In the CrossRuntime library, implement a very simple `MyCustomControl`:

```csharp
using System;
using Microsoft.UI.Xaml.Controls;

namespace UnoLibCrossRuntime
{
	public partial class MyCustomControl : UserControl
	{

#if WINDOWS
        const string Platform = "WINDOWS";
#elif __ANDROID__
        const string Platform = "ANDROID";
#elif __IOS__
        const string Platform = "__IOS__";
#elif __MACCATALYST__
        const string Platform = "MacCatylist";
#elif __SKIA__
        const string Platform = "SKIA";
#elif __WASM__
        const string Platform = "WASM";
#else
        const string Platform = "WTF";
#endif

        public MyCustomControl()
		{
            Content = new TextBlock()
                .Text(Platform);
		}
	}
}
```

4. refer to `UnoLib-CrossRuntime.csproj` in `UnoApp1.csproj`:
```xml
  <ItemGroup>
    <ProjectReference Include="..\UnoLibCrossRuntime\UnoLibCrossRuntime\UnoLibCrossRuntime.csproj" />
  </ItemGroup>
```

5. name the **StackPanel** in `UnoApp1.xaml`
```xaml
<Page x:Class="UnoApp1.MainPage"
      xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
      xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
      xmlns:local="using:UnoApp1"
      Background="{ThemeResource ApplicationPageBackgroundThemeBrush}">
  <StackPanel
        x:Name="_stackPanel"
        HorizontalAlignment="Center"
        VerticalAlignment="Center">
    <TextBlock AutomationProperties.AutomationId="HelloTextBlock"
          Text="Hello Uno Platform"
          HorizontalAlignment="Center" />
  </StackPanel>
</Page>

```
6. modify `UnoApp1.MainPage.xaml.cs`:
```csharp
namespace UnoApp1;

public sealed partial class MainPage : Page
{
    public MainPage()
    {
        this.InitializeComponent();
        _stackPanel.Children.Add(new UnoLibCrossRuntime.MyCustomControl());
    }
}
```

When I run in Windows:
<img width="604" alt="image" src="https://github.com/unoplatform/uno/assets/2528888/cc5a5b0b-925e-447f-80f5-2255d7731940">

But, when I run in WASM:
<img width="672" alt="image" src="https://github.com/unoplatform/uno/assets/2528888/bd3957c3-4b49-46fd-8fc9-514ef07aab95">

My question: is there a recommended approach to using a CrossRuntime library along side an Uno app within VisualStudio?

**Comment by jeromelaban** [comment]
Cross targeted libraries are generally meant to be used as nuget packages, mostly because the uno tooling will automatically perform a "bait and switch" at build time. When using such a project structure inside of Visual Studio using project references, this makes the swapping not working properly when not nuget packaged packaged.

We understand that platform compile-time specific content is difficult to build at this time for Wasm and Skia, mainly because we assumed that Microsoft would have been providing `net8.0-browser`, which did not happen.

We are exploring the ability to have a single-project structure, considering all the targets that uno provides, which would reduce this kind of friction.

At this point, we suggest that you use runtime conditions to discriminate between Wasm and Skia, similar to https://github.com/unoplatform/uno/discussions/11763#discussioncomment-5384614.

**Comment by baskren** [comment]
Thanks, Jerome, for clearing that all up.
